// gatsby-source-wordpress/dist/steps/create-schema-customization/helpers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.typeIsASupportedScalar = exports.typeIsABuiltInScalar = exports.introspectionFieldTypeToSDL = exports.getTypesThatImplementInterfaceType = exports.getTypeSettingsByType = exports.findTypeKind = exports.findNamedTypeName = exports.findNamedType = exports.filterTypeDefinition = exports.fieldOfTypeWasFetched = exports.buildTypeName = exports.buildInterfacesListForType = void 0;
var _merge2 = _interopRequireDefault(require("lodash/merge"));
var _cloneDeep2 = _interopRequireDefault(require("lodash/cloneDeep"));
var _store = _interopRequireDefault(require("../../store"));
var _typeFilters = require("./type-filters");
var _getGatsbyApi = require("../../utils/get-gatsby-api");
const buildInterfacesListForType = type => {
  let shouldAddNodeType = false;
  const list = ((type === null || type === void 0 ? void 0 : type.interfaces) || []).filter(interfaceType => {
    const interfaceTypeSettings = getTypeSettingsByType(interfaceType);
    return !interfaceTypeSettings.exclude && fieldOfTypeWasFetched(interfaceType);
  }).map(({
    name
  }) => {
    if (name === `Node`) {
      shouldAddNodeType = true;
    }
    return buildTypeName(name);
  });
  if (shouldAddNodeType) {
    list.push(`Node`);
  }
  return list;
};
exports.buildInterfacesListForType = buildInterfacesListForType;
let ceIntCache = null;
const isWpgqlOneThirteenZeroPlus = () => {
  if (ceIntCache !== null) {
    return ceIntCache;
  }
  const {
    typeMap
  } = _store.default.getState().remoteSchema;
  const connectionInterface = !!typeMap.get(`Connection`);
  const edgeInterface = !!typeMap.get(`Edge`);
  const isWpgqlOneThirteenZeroPlus = connectionInterface || edgeInterface;
  ceIntCache = isWpgqlOneThirteenZeroPlus;
  return isWpgqlOneThirteenZeroPlus;
};

/**
 * This function namespaces typenames with a prefix
 */
const buildTypeName = name => {
  if (!name || typeof name !== `string`) {
    return null;
  }
  const {
    schema: {
      typePrefix: prefix
    }
  } = (0, _getGatsbyApi.getPluginOptions)();

  // this is for our namespace type on the root { wp { ...fields } }
  if (name === prefix) {
    return name;
  }

  // Gatsby makes the same type, so we need to rename it to prevent conflicts
  if (name === `Filter`) {
    name = `FilterType`;
  }
  if (
  // Starting in WPGraphQL 1.13.0, Gatsby and WPGraphQL both generate types ending in these strings for every node type in the schema, so we need to rename types to prevent conflicts.
  // for users on older versions of WPGraphQL we should try to keep the schema how it was before
  isWpgqlOneThirteenZeroPlus() && (name.endsWith(`Connection`) || name.endsWith(`Edge`) || name.endsWith(`PageInfo`))) {
    name += `Type`;
  }
  if (name.startsWith(prefix)) {
    return name;
  }
  return prefix + name;
};

/**
 * Find the first type kind of a Type definition pulled via introspection
 * @param {object} type
 */
exports.buildTypeName = buildTypeName;
const findTypeKind = type => {
  if (type !== null && type !== void 0 && type.kind) {
    return type.kind;
  }
  if (type !== null && type !== void 0 && type.ofType) {
    return findTypeKind(type.ofType);
  }
  return null;
};
exports.findTypeKind = findTypeKind;
const findNamedType = type => {
  if (!type) {
    return null;
  }
  if (type.ofType) {
    return findNamedType(type.ofType);
  }
  return type;
};
exports.findNamedType = findNamedType;
const findNamedTypeName = type => {
  const namedType = findNamedType(type);
  return namedType === null || namedType === void 0 ? void 0 : namedType.name;
};
exports.findNamedTypeName = findNamedTypeName;
const fieldOfTypeWasFetched = type => {
  const {
    fetchedTypes
  } = _store.default.getState().remoteSchema;
  const typeName = findNamedTypeName(type);
  const typeWasFetched = !!fetchedTypes.get(typeName);
  return typeWasFetched;
};
exports.fieldOfTypeWasFetched = fieldOfTypeWasFetched;
const implementingTypeCache = new Map();
const getTypesThatImplementInterfaceType = type => {
  if (implementingTypeCache.has(type.name)) {
    return implementingTypeCache.get(type.name);
  }
  const state = _store.default.getState();
  const {
    typeMap
  } = state.remoteSchema;
  const allTypes = typeMap.values();
  const implementingTypes = Array.from(allTypes).filter(({
    interfaces
  }) => interfaces &&
  // find types that implement this interface type
  interfaces.find(singleInterface => singleInterface.name === type.name)).map(type => typeMap.get(type.name)).filter(type => type.kind !== `UNION` ||
  // if this is a union type, make sure the union type has one or more member types, otherwise schema customization will throw an error
  !!type.possibleTypes && !!type.possibleTypes.length);
  implementingTypeCache.set(type.name, implementingTypes);
  return implementingTypes;
};
exports.getTypesThatImplementInterfaceType = getTypesThatImplementInterfaceType;
const supportedScalars = [`Int`, `Float`, `String`, `Boolean`, `ID`, `Date`, `JSON`];
const typeIsABuiltInScalar = type =>
// @todo the next function and this one are redundant.
// see the next todo on how to fix the issue. If that todo is resolved, these functions will be identical. :(
supportedScalars.includes(findNamedTypeName(type));
exports.typeIsABuiltInScalar = typeIsABuiltInScalar;
const typeIsASupportedScalar = type => {
  if (findTypeKind(type) !== `SCALAR`) {
    // @todo returning true here seems wrong since a type that is not a scalar can't be a supported scalar... so there is some other logic elsewhere that is wrong
    // making this return false causes errors in the schema
    return true;
  }
  return supportedScalars.includes(findNamedTypeName(type));
};
exports.typeIsASupportedScalar = typeIsASupportedScalar;
const typeSettingCache = new Map();

// retrieves plugin settings for the provided type
const getTypeSettingsByType = type => {
  if (!type) {
    return {};
  }
  const typeName = findNamedTypeName(type);
  if (!typeName) {
    return {};
  }
  const cachedTypeSettings = typeSettingCache.get(typeName);
  if (cachedTypeSettings) {
    return cachedTypeSettings;
  }

  // the plugin options object containing every type setting
  const allTypeSettings = _store.default.getState().gatsbyApi.pluginOptions.type;
  const typeSettings = (0, _cloneDeep2.default)(allTypeSettings[typeName] || {});

  // the type.__all plugin option which is applied to every type setting
  const __allTypeSetting = (0, _cloneDeep2.default)(allTypeSettings.__all || {});
  if (typeName === `MediaItem`) {
    delete __allTypeSetting.limit;
    delete typeSettings.limit;
  }
  if (typeSettings) {
    const mergedSettings = (0, _merge2.default)(__allTypeSetting, typeSettings);
    typeSettingCache.set(typeName, mergedSettings);
    return mergedSettings;
  }
  typeSettingCache.set(typeName, __allTypeSetting);
  return __allTypeSetting;
};

/**
 * This is used to filter the automatically generated type definitions before they're added to the schema customization api.
 */
exports.getTypeSettingsByType = getTypeSettingsByType;
const filterTypeDefinition = (typeDefinition, typeBuilderApi, typeKind) => {
  const filters = _typeFilters.typeDefinitionFilters.filter(filter => [typeBuilderApi.type.name, `__all`].includes(filter.typeName));
  if (filters !== null && filters !== void 0 && filters.length) {
    filters.forEach(filter => {
      if (filter && typeof filter.typeDef === `function`) {
        typeDefinition = filter.typeDef(typeDefinition, typeBuilderApi, typeKind);
      }
    });
  }
  return typeDefinition;
};

// we should be using graphql-js for this kind of thing, but unfortunately this project didn't use it from the beginning so it would be a huge lift to refactor to use it now. In the future we will be rewriting this plugin using a new Gatsby source plugin toolkit, and at that time we'll use graphql-js.
// from introspection field types this will return a value like:
// `String` or `[String]` or `[String!]!` or `[String]!` or `[[String]]` or `[[String]!]!` or `[[String]!]`, etc
exports.filterTypeDefinition = filterTypeDefinition;
const introspectionFieldTypeToSDL = fieldType => {
  const openingTagsList = [];
  const closingTagsList = [];
  let reference = fieldType;
  while (reference) {
    switch (reference.kind) {
      case `SCALAR`:
        {
          const normalizedTypeName = supportedScalars.includes(reference.name) ? reference.name : `JSON`;
          openingTagsList.push(normalizedTypeName);
          break;
        }
      case `OBJECT`:
      case `INTERFACE`:
      case `UNION`:
        openingTagsList.push(buildTypeName(reference.name));
        break;
      case `NON_NULL`:
        closingTagsList.push(`!`);
        break;
      case `LIST`:
        openingTagsList.push(`[`);
        closingTagsList.push(`]`);
        break;
      default:
        break;
    }
    reference = reference.ofType;
  }
  return openingTagsList.join(``) + closingTagsList.reverse().join(``);
};
exports.introspectionFieldTypeToSDL = introspectionFieldTypeToSDL;
