// graphql-http/lib/audits/utils.mjs
/**
 *
 * audit/utils
 *
 */
export * from '../utils.mjs';
/**
 * Wrap and prepare an audit for testing.
 *
 * @private
 */
export function audit(name, fn) {
    return {
        name,
        fn: async () => {
            try {
                await fn();
                return {
                    name,
                    status: 'ok',
                };
            }
            catch (errOrReason) {
                if (typeof errOrReason !== 'string') {
                    // anything thrown that is not an assertion string is considered fatal
                    throw errOrReason;
                }
                return {
                    name,
                    status: name.startsWith('MUST')
                        ? // only failing MUSTs are considered errors
                            'error'
                        : // everything else is optional and considered a warning
                            'warn',
                    reason: errOrReason,
                };
            }
        },
    };
}
/**
 * Will throw a string if the assertion fails.
 *
 * All fatal problems will throw an instance of Error.
 *
 * @private
 */
export function assert(name, actual) {
    return {
        toBe: (expected) => {
            if (actual !== expected) {
                throw `${name} ${actual} is not ${expected}`;
            }
        },
        toBeBetween: (min, max) => {
            if (!(min <= actual && actual <= max)) {
                throw `${name} ${actual} is not between ${min} and ${max}`;
            }
        },
        toContain: (expected) => {
            // @ts-expect-error types will match, otherwise never
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            if (!actual.includes(expected)) {
                throw `${name} ${JSON.stringify(actual)} does not contain ${JSON.stringify(expected)}`;
            }
        },
        notToContain: (expected) => {
            // @ts-expect-error types will match, otherwise never
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            if (actual.includes(expected)) {
                throw `${name} ${JSON.stringify(actual)} contains ${JSON.stringify(expected)}`;
            }
        },
        toHaveProperty: (
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        prop) => {
            // @ts-expect-error types will match, otherwise never
            if (!(prop in actual)) {
                throw `${name} ${JSON.stringify(actual)} does not have a property '${String(prop)}'`;
            }
        },
        notToHaveProperty: (
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        prop) => {
            // @ts-expect-error types will match, otherwise never
            if (prop in actual) {
                throw `${name} ${JSON.stringify(actual)} does have a property '${String(prop)}'`;
            }
        },
    };
}
/**
 * Parses the string as JSON and safely reports parsing issues for audits.
 *
 * Assumes the parsed JSON will be an `ExecutionResult`.
 *
 * @private
 * */
export async function assertBodyAsExecutionResult(res) {
    const str = await res.text();
    try {
        return JSON.parse(str);
    }
    catch (err) {
        throw `Response body is not valid JSON. Got ${JSON.stringify(str)}`;
    }
}
